compiler/rustc_hir_typeck/src/fn_ctxt/adjust_fulfillment_errors.rs

```diff
@@ -714,12 +714,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     self.tcx.parent(expr_ctor_def_id)
                 }
                 hir::def::DefKind::Ctor(hir::def::CtorOf::Variant, hir::def::CtorKind::Fn) => {
-                    // If this is a variant, its parent is the type definition.
-                    if in_ty_adt.did() != self.tcx.parent(expr_ctor_def_id) {
-                        // FIXME: Deal with type aliases?
+                    if in_ty_adt.did() == self.tcx.parent(expr_ctor_def_id) {
+                        // The constructor definition refers to the variant:
+                        // For example, for a local type `MyEnum::MyVariant` triggers this case.
+                        expr_ctor_def_id
+                    } else if in_ty_adt.did() == self.tcx.parent(self.tcx.parent(expr_ctor_def_id))
+                    {
+                        // The constructor definition refers to the "constructor" of the variant:
+                        // For example, `Some(5)` triggers this case.
```

Comment:
I'm not sure I understand... Does `Some`'s def id refer to the `use` statement?
```rust
// (3) and then parent's parent refers to the option itself
enum Option<T> {
    Some(T), // (2) The first parent refers here
    None,
}

pub use Option::Some; // (1) `Some`'s def id refers to here
```

If this is the case, would this "fail" on something like the following?:
```rust
enum A<T> { V0(T) }
use A::V0 as V1;
use V1 as V2;
/* then, code with V2, such that you need parent's parent's parent */
```

Reply:
It's not pointing to the `use` statement. It's pointing to a "constructor" definition which lives inside the "variant" definition.

So if we have

```rs
enum ExampleTuple<T> {
    ExampleTupleVariant(T),
}
use ExampleDifferentTupleVariantName as ExampleYetAnotherTupleVariantName;
use ExampleTuple as ExampleOtherTuple;
use ExampleTuple::ExampleTupleVariant as ExampleDifferentTupleVariantName;
use ExampleTuple::*;

impl<A> T1 for ExampleTuple<A> where A: T3 {}
```

then for all of

- `ExampleTuple::ExampleTupleVariant(q)`
- `ExampleTupleVariant(q)`
- `ExampleOtherTuple::ExampleTupleVariant(q)`
- `ExampleDifferentTupleVariantName(q)`
- `ExampleYetAnotherTupleVariantName(q)`

by printing the relevant `DefId`s we see the same thing in each case:

```
expr_ctor_def_id :::                                   DefId(0:29 ~ blame_trait_error[b442]::ExampleTuple::ExampleTupleVariant::{constructor#0})
self.tcx.parent(expr_ctor_def_id) :::                  DefId(0:28 ~ blame_trait_error[b442]::ExampleTuple::ExampleTupleVariant)
self.tcx.parent(self.tcx.parent(expr_ctor_def_id)) ::: DefId(0:26 ~ blame_trait_error[b442]::ExampleTuple)
```

So going up once gets us to the variant and going up twice gets us to the type definition. I'm actually not sure how/why going up only once ever seemed to work before - this may have just been a miss since the first PR focused mostly on structs and tuples.

I've added a lot of new tests to cover this name resolving behavior.

Reply:
Ohhh, and for structures you get `expr_ctor_def_id ::: DefId(...::Struct::{constructor#0})`, so you only need to go one level up the parent chain. Makes sense, thanks for the explanation and adding a comment!